### PR TITLE
Adds test to ensure custom contexts are usable

### PR DESCRIPTION
**Pull request by hafeidejiangyou**
I want to hold the my Context and use it in the host funtion call, So I add a holder funtion to hold the Context instance.
I think this is a feature, Please review it.

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
Hi! because context is already propagated, you can use normal context keys instead to add state for host function calls. That's how we do it internally already for host imports like wasi.


First define the context state you need.
```go
// requestStateKey is a context.Context value associated with a requestState
// pointer to the current request.
type requestStateKey struct{}

type requestState struct {
  magic uint64
}
```

Next, define your host function using this state
```go
func getMagic(ctx context.Context) uint64 {
	return requestStateFromContext(ctx).magic
}

func requestStateFromContext(ctx context.Context) *requestState {
	return ctx.Value(requestStateKey{}).(*requestState)
}
```

Finally, when something calls in, pass the same context
```go
	 ctx := context.WithValue(context.Background(), requestStateKey{}, s)
     	_, err = mod.ExportedFunction("get_magic").Call(ctx)
```

**Comment by hafeidejiangyou** [comment]
I had try, benchmark show use context.Context is so slow in host function call, so I change it.

**Comment by codefromthecrypt** [comment]
@hafeidejiangyou I think if you want a faster way, you can find a faster one or make your own context. We don't decorate the context internally in any way.

Ex. Define a custom context.
```
var _ context.Context = &MyContext{}

type MyContext struct {
	magic uint64
}

func (c *MyContext) Deadline() (deadline time.Time, ok bool) {
	return
}

func (c *MyContext) Err() (err error) {
	return 
}

func (c *MyContext) Value(any)(val any) {
	return
}

func (f *fastCtx) Done() (done <-chan struct{}) {
	return
}
```

Use it directly in your functions:
```
func getMagic(ctx context.Context) uint64 {
	return ctx.(*MyContext).magic
}
```

I think this is the best way to improve performance vs making a new type to pass around.

**Comment by hafeidejiangyou** [comment]
OK, I agree it is a better way. 
Does the context propagate will be intended for other uses when `ExportedFunction.Call(ctx, args)` is called？ I mean the `ctx` passed to `Call` can not be decorate like `newCtx = context.WithValue(ctx, "hello", "World")`, because it will panic with cast `ctx.(*MyContext)` in host function

**Comment by codefromthecrypt** [comment]
@hafeidejiangyou we currently don't decorate it, but of course if you are making a shared library and they decorate externally that is another issue. If you want to make sure this works, maybe change your PR to a runtime_test that does exactly this. this way, it will break our build if any problem later.

**Comment by hafeidejiangyou** [comment]
Ok, I will add a test for it.

File: runtime_test.go [hunk 1]
```diff
@@ -533,6 +549,75 @@ func TestRuntime_CloseWithExitCode(t *testing.T) {
 	}
 }
 
+func TestCallHostWithSelfImplContext(t *testing.T) {
```
Comment (codefromthecrypt) [inline runtime_test.go:552]:
```suggestion
func TestHostFunctionWithCustomContext(t *testing.T) {
```

**Comment by codefromthecrypt** [comment]
thanks for this!

**Pull request merged**